felix/dataplane/linux/route_mgr.go

```diff
@@ -0,0 +1,686 @@
+			logrus.Debug("Sleep returning early: context finished.")
+		case <-m.tunnelChangedC:
+			logrus.Debug("Sleep returning early: tunnel changed.")
+		}
+	}
+
+	for ctx.Err() == nil {
+		if m.dataIfaceAddr() == "" {
+			m.logCtx.Debug("Missing local information, retrying...")
+			sleepMonitoringChans(10 * time.Second)
+			continue
+		}
+
+		dataDevice, err := m.detectDataIface()
+		if err != nil {
+			m.logCtx.WithError(err).Warn("Failed to find data device, retrying...")
+			sleepMonitoringChans(1 * time.Second)
+			continue
+		}
+
+		link, addr, err := getDevice(dataDevice)
+		if err != nil {
+			m.logCtx.WithError(err).Warn("Failed to find data device, retrying...")
+			sleepMonitoringChans(1 * time.Second)
+			continue
+		}
+
+		m.logCtx.Debug("Configuring tunnel device")
+		err = m.configureTunnelDevice(link, addr, mtu, xsumBroken)
+		if err != nil {
+			m.logCtx.WithError(err).Warn("Failed to configure tunnel device, retrying...")
+			logNextSuccess = true
+			sleepMonitoringChans(1 * time.Second)
+			continue
+		}
+
+		newDataIface := dataDevice.Attrs().Name
+		if newDataIface != dataIface {
+			// Send a message back to the main loop to tell it to update the
+			// routing tables.
+			m.logCtx.Infof("data device changed from %q to %q", dataIface, newDataIface)
+			select {
+			case dataIfaceC <- newDataIface:
+				dataIface = newDataIface
+			case <-m.tunnelChangedC:
+				m.logCtx.Info("Tunnel changed; restarting configuration.")
+				continue
+			case <-ctx.Done():
+				continue
+			}
+		}
+
+		if logNextSuccess {
+			m.logCtx.Info("Tunnel device configured")
+			logNextSuccess = false
+		}
+		sleepMonitoringChans(wait)
+	}
+}
+
+func (m *routeManager) configureTunnelDevice(
+	newLink netlink.Link,
+	addr string,
+	mtu int, xsumBroken bool) error {
+	if newLink == nil {
+		return fmt.Errorf("no tunnel link provided")
+	}
+	logCtx := m.logCtx.WithFields(logrus.Fields{
+		"mtu":        mtu,
+		"tunnelAddr": addr,
+	})
+	logCtx.Debug("Configuring tunnel device")
+	// Try to get the device.
+	link, err := m.nlHandle.LinkByName(m.tunnelDevice)
+	if err != nil {
+		m.logCtx.WithError(err).Info("Failed to get tunnel device, assuming it isn't present")
+		if m.maintainIPOnly {
+			return err
+		}
+		if err := m.nlHandle.LinkAdd(newLink); err == syscall.EEXIST {
+			// Device already exists - likely a race.
+			m.logCtx.Debug("Tunnel device already exists, likely created by someone else.")
+		} else if err != nil {
+			// Error other than "device exists" - return it.
+			return err
+		}
+
+		// The device now exists - requery it to check that the link exists and is our tunnel device.
+		link, err = m.nlHandle.LinkByName(m.tunnelDevice)
+		if err != nil {
+			return fmt.Errorf("can't locate created tunnel device %v", m.tunnelDevice)
+		}
+	}
+
+	if m.maintainIPOnly {
+		if err := m.ensureAddressOnLink(addr, link); err != nil {
+			return fmt.Errorf("failed to ensure address of interface: %s", err)
+		}
+		return nil
+	}
+
+	if m.vxlanEnabled() {
```

Comment:
why did you put this whole "keep device i sync" here as well? It seems like there quite a bit of a difference betwee the two tunnels. With this it also feel more like a tunnelBaseMgr then just route mgr. It should probably be called tunnelRouteMgr as it ia quite a specific type of route manager

Reply:
Actually there is not much difference. The Only difference is in the link itself, otherwise all other operation related to making sure the device is created, and configured is the same.

Reply:
In the end, route management is tightly integrated with interface management. I did another prototype to make them separate, but the result was not clean. Both ipip and vxlan managers have a keep in sync and tunnel configuration methods with around 90% code duplicate. This is the reason, I moved it into this manager.

This check is the only place that we have to check specifically for vxlan. Though this one can also be improved to be generic.

We can definitely change the names as we think appropriate.

Reply:
btw, I initially named this manager `tunnelManager`, but then changed it to `routeManager` to use it later for unencap routing.